staging/src/k8s.io/apiserver/pkg/storage/etcd3/etcdfeature/feature_support_checker.go

```diff
@@ -0,0 +1,147 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package etcdfeature
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	clientv3 "go.etcd.io/etcd/client/v3"
+	"k8s.io/apimachinery/pkg/util/version"
+	"k8s.io/apiserver/pkg/storage"
+	"k8s.io/klog/v2"
+)
+
+var (
+	// Define these static versions to use for checking version of etcd, issue on kubernetes #123192
+	v3_4_25 = version.MustParseSemantic("3.4.25")
+	v3_5_0  = version.MustParseSemantic("3.5.0")
+	v3_5_8  = version.MustParseSemantic("3.5.8")
+
+	// DefaultFeatureSupportChecker is a shared global FeatureSupportChecker.
+	DefaultFeatureSupportChecker FeatureSupportChecker = newDefaultFeatureSupportChecker()
+)
+
+// FeatureSupportChecker to define Supports functions.
+type FeatureSupportChecker interface {
+	Supports(feature storage.Feature) (bool, error)
+	// This method works with etcd client to recalcuate feature support and caches it.
+	EndpointSupports(ctx context.Context, c client, endpoint string, feature storage.Feature) (bool, error)
+	// Only use for testing and use defer afterwork.
+	SetFeatureSupportCheckerDuringTest(feature storage.Feature, value bool) func()
+}
+
+// featureSupportInfo holds information about feature support status, used for cache.
+type featureSupportInfo struct {
+	Supported bool
+}
+
+type defaultFeatureSupportChecker struct {
+	progresNotifySupport sync.Map
+}
```

Comment:
```suggestion
type defaultFeatureSupportChecker struct {
	mux sync.Mutex
	progressNotifySupported bool
	progresNotifyEndpointCache map[string]bool
}
```

Reply:
Let's adjust semantics of the results returned by `defaultFeatureSupportChecker`,
* `Supports(feature storage.Feature) (bool, error)` returns:
  * `true, nil` - if empty or all of the endpoints support the feature.
  * `false, nil` - if any endpoint doesn't support the feature
  * `false, err` - if feature is not implemented.
* `ClientSupports(ctx context.Context, c client, feature storage.Feature) (bool, error)` returns
  * `true, nil` - if all endpoints previously encountered and those provided by client support the feature.
  * `false, nil` - if any endpoints previously encountered or those provided by client don't support the feature.
  * `false, err` - if we failed checking if client supports the feature.

Reply:
**if empty** or all of the endpoints support the feature. **Or the cache is empty**

`Or the cache is empty` is already included

Reply:
The difference is 
**if all endpoints previously encountered ** and those provided by client support the feature.

Previously we only counted endpoints that were provided by client.

Reply:
My bad, you're right.
> if empty or all of the endpoints support the feature. Or the cache is empty

OK i will check them out in a bit.

Reply:
Just one question, if we want to add `progressNotifySupported bool` i think we should make it `true` at `newDefaultFeatureSupportChecker`  and return `progressNotifySupported` at calling `Supports` and only update it when `ClientSupports` gets new endpoints and one of endpoints **doesn't support** or **having trouble checking the version** right?

Reply:
So in that case we don't need to cache all endpoints right?

Reply:
Right